[generate] Run custom generation code from the Hub (#36405)
huggingface/transformers: docs/source/en/generation_strategies.md

This guide will help you understand the different decoding strategies available in Transformers and how and when to use them.

## Basic decoding methods

> These are well established decoding methods, and should be you starting point for text generation tasks.

Reviewer: ```suggestion
These are well established decoding methods, and should be your starting point for text generation tasks.
```

(typo)

This guide will help you understand the different decoding strategies available in Transformers and how and when to use them.

> [!TIP]
> `transformers` can also load custom generation strategies from the Hub! See the `Custom decoding methods` section below for instructions on how to load or share a custom generation strategy.

> ## Basic decoding methods

Author: Reorganized the generation strategies docs to make it easier to find the docs regarding custom generation methods.

The outer level is now `basic decoding methods` (moved greedy decoding, sampling, and beam search here), `advanced decoding methods` (moved the other ones here), and `custom decoding methods` (new section for the code on the Hub)

- have the model continue thinking if it is uncertain;
- roll back generation if the model gets stuck;
- handle special tokens with custom logic;
- enhanced input preparation for advanced models;
- (...)

To that end, we enable `generate` to load custom generation recipes from the Hub through its `recipe` argument. This means anyone can create and share their own advanced generation method which, from a user perspective, requires no additional python packages.

<!-- TODO before merging: 1) better repo name (use a `generate-community` org?) 2) prettify the repo -->
```py
from transformers import AutoModelForCausalLM, AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2.5-0.5B-Instruct")
model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2.5-0.5B-Instruct", device_map="auto")

inputs = tokenizer(["The quick brown"], return_tensors="pt").to(model.device)
gen_out = model.generate(**inputs, recipe="joaogante/test_generate_from_hub")  # minimal greedy decoding
print(tokenizer.batch_decode(gen_out, skip_special_tokens=True)[0])
'The quick brown fox jumps over a lazy dog, and the dog is a type of animal. Is'
```

> ### Using a custom decoding method
> 
> After you've identified a custom generation recipe on the Hub that you'd like to try, the first thing you should do is to read its README page. When creating a recipe (see below), we recommend their creators to document new arguments and output type differences, if any of these exist. Reasonably, you should assume everything works like the base `generate` unless documented.

Reviewer 2: ```suggestion
You should read the `README.md` file of a custom generation strategy to see what the new arguments and output type differences are, if they exist. Otherwise, you can assume it works like the base [`~GenerationMixin.generate`] method.
```

> Let's consider the Hub repository [`joaogante/test_generate_from_hub`](https://huggingface.co/joaogante/test_generate_from_hub). We can see in its README that it has an additional input argument, `left_padding`, which is an optional integer that will add that number of pad tokens before the prompt. Let's try it out!

Reviewer 2: ```suggestion
Consider the Hub repository [joaogante/test_generate_from_hub](https://huggingface.co/joaogante/test_generate_from_hub) as an example. The `README.md` states that it has an additional input argument, `left_padding`, which adds a number of padding tokens before the prompt.
```

```py
gen_out = model.generate(**inputs, recipe="joaogante/test_generate_from_hub", left_padding=5)
print(tokenizer.batch_decode(gen_out)[0])
'<|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|>The quick brown fox jumps over the lazy dog.\n\nThe sentence "The quick'
```

> If the recipe has pinned python requirements and your environment doesn't fulfill them, you'll get an exception about missing requirements. For instance, [`joaogante/test_generate_from_hub_2`](https://huggingface.co/joaogante/test_generate_from_hub_2) has an impossible set of requirements defined in its `requirements.txt`, and you'll see something like this if you try to run it:

Reviewer 2: ```suggestion
If the recipe has pinned Python requirements that your environment doesn't meet, you'll get an exception about missing requirements. For instance, [joaogante/test_generate_from_hub_2](https://huggingface.co/joaogante/test_generate_from_hub_2) has an impossible set of requirements defined in its `requirements.txt`, and you'll see the error message below if you try to run it.
```

```
ValueError: Missing requirements for joaogante/test_generate_from_hub_2:
foo (installed: None)
bar==0.0.0 (installed: None)
torch>=99.0 (installed: 2.6.0+cu126)
```

> Updating your python requirements accordingly will remove this error message.

Reviewer 2: ```suggestion
Updating your Python requirements accordingly will remove this error message.
```

### Creating a custom decoding method

> To create a new decoding method, you need to create a new "Model" repository and push a few files into it:

Reviewer 2: ```suggestion
To create a new decoding method, you need to create a new [**Model**](https://huggingface.co/new) repository and push a few files into it.
```

> 1. (required) `generate.py`, which contains all the logic for your custom decoding method.
> 2. (optional) `README.md`, the front page on the Hub to your custom generation method;
> 3. (optional) `requirements.txt`, to add new python requirements and/or lock specific versions for correct utilization of the technique.
> 
> Let's dive at each file, and have a look at their requirements and best practices.

Reviewer 2: ```suggestion
1. `generate.py`, which contains all the logic for your custom decoding method.
2. `README.md`, you should document any new arguments or output type differences of your custom method here.
3. `requirements.txt`, used to add new Python requirements and/or lock specific versions to correctly use a technique.
```

#### generate.py

> This is the core of your decoding method. It *must* contain a `generate` method, and this method *must* contain a `model` argument as its first argument. `model` is the model instance, which means you have access to all attributes and methods in the model, including the ones defined in `GenerationMixin` (like the original `generate` method).

Reviewer 2: ```suggestion
This is the core of your decoding method. It *must* contain a [`~GenerationMixin.generate`] method, and this method *must* contain a `model` argument as its first argument. `model` is the model instance, which means you have access to all attributes and methods in the model, including the ones defined in [`GenerationMixin`] (like the base `generate` method).
```

> Under the hood, when the original `generate` method is called with a `recipe` argument, it will first check its requirements (if they exist, see below), then locate the custom `generate` method in `generate.py`, and finally calls the custom `generate`, forwarding all received arguments to it (plus `model`). This means your `generate` can have a mix of orginal and custom arguments, as well as a different output type -- the world is your oyster!
> 
> Here is an example of a `generate.py` file that mixes original and custom arguments:

Reviewer 2: ```suggestion
Under the hood, when the base [`~GenerationMixin.generate`] method is called with a `recipe` argument, it first checks its Python requirements, then locates the custom `generate` method in `generate.py`, and finally calls the custom `generate`. All received arguments and `model` are forwarded to the custom `generate` method.

This means your `generate` can have a mix of original and custom arguments (as well as a different output type) as shown below.
```

    max_length = generation_config.max_length or cur_length + generation_config.max_new_tokens

    # Example of custom argument: add `left_padding` (integer) pad tokens before the prompt
    if left_padding is not None:
        if not isinstance(left_padding, int) or left_padding < 0:
            raise ValueError(f"left_padding must be an integer larger than 0, but is {left_padding}")

        pad_token = kwargs.pop("pad_token", None) or generation_config.pad_token_id or model.config.pad_token_id
        if pad_token is None:
            raise ValueError("pad_token is not defined")
        batch_size = input_ids.shape[0]
        pad_tensor = torch.full(size=(batch_size, left_padding), fill_value=pad_token).to(input_ids.device)
        input_ids = torch.cat((pad_tensor, input_ids), dim=1)
        cur_length = input_ids.shape[1]

    # Simple greedy decoding loop
    while cur_length < max_length:
        logits = model(input_ids).logits
        next_token_logits = logits[:, -1, :]
        next_tokens = torch.argmax(next_token_logits, dim=-1)
        input_ids = torch.cat((input_ids, next_tokens[:, None]), dim=-1)
        cur_length += 1

    return input_ids
```

> Recommended practices:

Reviewer 2: ```suggestion
Follow the recommended practices below to ensure your custom method works as expected.
```

> - The original `generate` has extensive logic for validation and input preparation. Feel free to reuse them;
> - If you use any private method/attribute in `model`, pin the `transformers` version you're using in the requirements;
> - Consider adding model validation, input validation, or even a separate test file to help users sanity-check your code in their environment.

Reviewer 2: ```suggestion
- Feel free to reuse the logic for validation and input preparation in the original [`~GenerationMixin.generate`].
- Pin the Transformers version in the requirements if you use any private method/attribute in `model`.
- Consider adding model validation, input validation, or even a separate test file to help users sanity-check your code in their environment.
```

#### README.md

> As the front page of your decoding method, this is where other users get to know it. In addition to a description of the method, we heavily recommend documenting any input and/or output differences to the original `generate`. That way, users can focus on what's new, and rely on `transformers` docs for generic implementation details.
> 
> Recommended pratices:
> - Document input and output differences in `generate`;
> - Add self-contained examples to enable quick experimentation;
> - Describe additional soft-requirements (e.g. if the technique only works well with a certain family of models).

Reviewer 2: ```suggestion
The `README.md` describes the decoding method to other users. In addition to a description of the method, we heavily recommend documenting any input and/or output differences to the original [`~GenerationMixin.generate`]. This way, users can focus on what's new, and rely on Transformers docs for generic implementation details.

Some more recommended practices are listed below to help users get familiar with your decoding method.
- Document input and output differences in [`~GenerationMixin.generate`].
- Add self-contained examples to enable quick experimentation.
- Describe soft-requirements such as if the method only works well with a certain family of models.
```

#### requirements.txt

> If your decoding method has python requirements that are not present in the original `transformers` package, you can specify then in a `requirements.txt` file (just like you would in a python project). These will be checked at runtime, and an exception will be thrown if they are not honored.
> 

Reviewer 2: ```suggestion
Specify additional Python requirements in a `requirements.txt` file for your decoding method if they're not present in Transformers. These are checked at runtime and an exception will be thrown if they're missing.
```

Enable diverse beam search with the `num_beams`, `num_beam_groups` and `diversity_penalty` parameters (the `num_beams` parameter should be divisible by `num_beam_groups`).

```py
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("meta-llama/Llama-2-7b-hf")
inputs = tokenizer("Hugging Face is an open-source company", return_tensors="pt").to("cuda")

model = AutoModelForCausalLM.from_pretrained("meta-llama/Llama-2-7b-hf", torch_dtype=torch.float16).to("cuda")
# explicitly set to 100 because Llama2 generation length is 4096
outputs = model.generate(**inputs, max_new_tokens=50, num_beams=6, num_beam_groups=3, diversity_penalty=1.0, do_sample=False)
tokenizer.batch_decode(outputs, skip_special_tokens=True)
'Hugging Face is an open-source company 🤗\nWe are an open-source company. Our mission is to democratize AI and make it accessible to everyone. We believe that AI should be used for the benefit of humanity, not for the benefit of a'
```

> ## Custom decoding methods

Author: The new docs start here

```py
from transformers import AutoModelForCausalLM, AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2.5-0.5B-Instruct")
model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2.5-0.5B-Instruct", device_map="auto")

inputs = tokenizer(["The quick brown"], return_tensors="pt").to(model.device)
# `custom_generate` replaces the original `generate` by the custom decoding method defined in
# `transformers-community/custom_generate_example`
gen_out = model.generate(**inputs, custom_generate="transformers-community/custom_generate_example", trust_remote_code=True)
print(tokenizer.batch_decode(gen_out, skip_special_tokens=True)[0])
'The quick brown fox jumps over a lazy dog, and the dog is a type of animal. Is'
```

You should read the `README.md` file of the repository containing the custom generation strategy to see what the new arguments and output type differences are, if they exist. Otherwise, you can assume it works like the base [`~GenerationMixin.generate`] method.

> [!TIP]

> > You can find all custom decoding methods by [searching for their custom tag](https://huggingface.co/models?other=custom_generate), `custom_generate`

Author: Good idea 👍 Will add
